packages/openapi-v3/src/__tests__/unit/spec-contributor/fixtures/security.spec.extension.ts

```diff
@@ -0,0 +1,37 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/openapi-v3
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {bind} from '@loopback/context';
+import {ReferenceObject, SecuritySchemeObject} from '../../../../';
+import {
+  asSpecEnhancer,
+  OAISpecEnhancer,
+} from '../../../../extension-point/types';
+import {OpenApiSpec} from '../../../../types';
+
+export type SecuritySchemeObjects = {
+  [securityScheme: string]: SecuritySchemeObject | ReferenceObject;
+};
+
+export const SECURITY_SCHEME_SPEC: SecuritySchemeObjects = {
+  bearerAuth: {
+    type: 'http',
+    scheme: 'bearer',
+    bearerFormat: 'JWT',
+  },
+};
+
+/**
+ * A spec contributor to add bearer token
+ * OpenAPI security entry
+ */
+@bind(asSpecEnhancer)
+export class SecuritySpecEnhancer implements OAISpecEnhancer {
+  modifySpec(spec: OpenApiSpec) {
+    spec.components = spec.components ?? {};
+    spec.components.securitySchemes = SECURITY_SCHEME_SPEC;
+    return spec;
```

Comment:
I am little bit concerned about the design where the spec-enhancer is modifying the spec directly. Every implementation has to deal with detecting whether optional objects like `spec.compontents` are set and be prepared to handle the case when they are not present. 

Have you considered a different approach, where  the enhancer provides new data to be added to the spec, and LB runtime ensures the data is cleverly merged together? 

Take the `InfoSpecEnhancer` for an example. IIUC, it will discard any `info` fields provided by the app developer and replace them with `info` object containing only `title` and `version`. This problem is easy to miss when reviewing the code. I would prefer to design spec-enhancing infrastructure in a way that will make errors like these more difficult to make.

Reply:
> Have you considered a different approach, where the enhancer provides new data to be added to the spec, and LB runtime ensures the data is cleverly merged together?

Good question, I understand your concern and the risk to give enhancers the right to modify the entire spec, and actually for the original issue(add endpoint level security spec for auth automatically) we are addressing, WE CONTROL THE MERGING LOGIC is easier than my current approach in this PR.

While my concern is...for a fragment contributed by an arbitrary extension, how do we decide the merging logic?

Reply:
Add endpoint level security spec as follow:
```
paths:
  /something:
    get:
      security:
        - basicAuth:[]
```
- Developer decorate endpoint as `@authenticate('basicAuth')`
- Basic auth strategy component provide the security spec fragment and register the it as an OASEnhancer extension.(here the extension only provides the spec, but don't have merge logic)
- In the rest module, we first generate the entire `paths` spec for the controller methods with auth metadata included, e.g. `x-auth-strategy-name: 'basicAuth'`. Then iterate through all the endpoints, retrieve the security spec fragment based on their auth metadata by calling `OASenhancer.getSpecByName('basicAuth')`, merge fragment into the path spec.

Reply:
> While my concern is...for a fragment contributed by an arbitrary extension, how do we decide the merging logic?

I am proposing to start with the simplest option that works for security schemas. We can look into more flexible & powerful merging strategies later, as we learn about new different use cases.

I think the initial merging strategy can do the following:
- If the value is scalar (not object, not an array), then set it as-is.
- If the value is an object, then recursively merge all properties.
- If the value is an array, then append items provided by OASenhancer to the existing array.

A mock-up implementation:

```js
function merge(current, enhancement) {
  const result = _.cloneDeep(current);
  for (key in enhancement) {
    const val = enhancement[key];
    if (Array.isArray(val)) {
      // TODO: verify that current[key] is either `undefined` or an array
      result[key] = [...current[key], ...val];
    } else if (typeof val === 'object') {
      // TODO: verify that current[key] is either `undefined` or an object
      result[key] = merge(current[key] ?? {}, val);
    } else {
      result[key] = val;
    }
  }
  return result;
}
```

Thoughts?

Reply:
Ah, based on our chat, I see that I misunderstood this thread. My comment above is proposing a merging rule for [security schemes](https://github.com/OAI/OpenAPI-Specification/blob/3.0.1/versions/3.0.1.md#componentsSecuritySchemes) in `/components/security`, not for endpoint-level [security requirements](https://github.com/OAI/OpenAPI-Specification/blob/3.0.1/versions/3.0.1.md#security-requirement-object).

**When it comes to security requirements at endpoint level, I feel it's important to have a somewhat generic tool provided by `@loopback/rest`, a tool that any custom authentication solution can leverage.**

This will make it easy for any extensions to contribute endpoint-level security requirements. If somebody decides to build an alternative to `@loopback/authentication` extension, then they should have all tools available and should not need to re-implement contributions of endpoint security requirements.

The rest is just implementation details in a sense.

---

One option is to use the spec enhancer as proposed so far.

In https://github.com/strongloop/loopback-next/pull/4258/files#r357211992, I proposed to implement a helper function for merging OpenAPI spec fragments. If we decide to use the concept of OpenAPI spec enhancer as the tool for contributing endpoint security requirements, then I would like `@loopback/rest` to expose a function for setting (or adding) a new security requirement for the given endpoints. 

For example (assuming a hard-coded security spec to keep the example simpler):

```ts
import {assignSecurityRequirement, /*...*/} from '@loopback/rest';

@bind(asSpecEnhancer)
export class BasicAuthSpecEnhancer implements OAISpecEnhancer {
  modifySpec(spec: OpenApiSpec) {
    return assignSecurityRequirement({
      spec, // the original spec
      verb: 'get', // http verb
      path: '/greet',  // http path
      name: 'petstore_auth', // name of the security schema
      scopes: ['read:pets'], // optional, defaults to []
   });
  }
}
```

On the second thought, maybe we don't need any new function and `assignSpecFragment` is actually all we need:

```ts
assignSpecFragment(spec, {
  paths: {
    [endpointPath]: {
      [endpointVerb]: {
        security: {
          petstore_auth: [
            'read:pets'
          ],
         }
       }
     }
  }
});
```

---

Another option is to introduce a new decorator. If you think about how we are defining endpoint metadata now, we have several decorators like `@param` and `@requestBody` that are contributing spec fragments that are later combined into a single operation object. I have an idea to implement a new decorator for defining security requirements.

Example usage:

```ts
import {get, security} from '@loopback/rest'

class GreetController {
  @security('petstore_auth', [
    "write:pets",
    "read:pets",
  ])
  @get('/hello')
   greet() {
     return 'hello world';
  }
}
```

The `@authenticate` decorator can then call `@security` under the hood to contribute the correct security requirement to the OpenAPI spec.

```ts
import {get} from '@loopback/rest';
import {authenticate} from '@loopbakc/authenticate';

class GreetController {
  @authenticate('basicAuth')
  // ^-- will call security('basicAuth') under the hood
  @get('/hello')
   greet() {
     return 'hello world';
  }
}
```

Anyhow. I don't know if the `@security` decorator is a viable solution that can support our current auth/auth design.


I am posting these ideas just for inspiration, please don't hold to them.

Reply:
@bajtos Good summary, based on all our discussions, I changed the implementation and design to be:

- I still kept the `modifySpec` function in the interface to give people the max flexibility to modify the original spec:
- As we chatted in the meeting, I introduced a default spec merge function(merge a patch spec into the original spec), and exported it for extension developers to use in the `modifySpec` function.
  - It leverages an existing `json-merge-patch` module to merge the spec. I chose it because [`ajv` module](https://github.com/epoberezkin/ajv-merge-patch) also uses it. More details can be found in the API documentation.
- I added two more util functions to the service `getEnhancerByName` and `applyEnhancerByName`, so people can apply a single enhancer in case they need, or doesn't  want to **automatically** apply all the enhancers in `generateSpec`.
- For the endpoint level security spec, I tend to agree with adding a new decorator `@security()`. Reasons are same as what you explained:
  - "When it comes to security requirements at endpoint level, I feel it's important to have a somewhat generic tool provided by @loopback/rest, a tool that any custom authentication solution can leverage."